public/static/docs/api-reference/open.md

```diff
@@ -0,0 +1,171 @@
+
+> See also `dvc.api.read()` for a shorthand way to read the complete contents of
+> a tracked file.
+
+💡 Note that `dvc.api.open()` is able to
+[stream](https://docs.python.org/3/library/io.html) the file directly from
+**most**
+[remote types](/doc/command-reference/remote/add#supported-storage-types)
+(local, S3, Azure, GCP, OSS, SSH, HDFS). Otherwise, the file is downloaded
+regularly into a temporary local path before the file object is made available.
+
+> This has similar uses as the `dvc get` and `dvc import` CLI commands.
+
+## Parameters
+
+- **`path`** - specifies the location of the target data within the source
+  project in `repo`, relative to the project's root.
+
+- `repo` - specifies the location of the source DVC project. If not supplied,
+  defaults to the current DVC project. It can be a URL or a file system path.
+  Both HTTP and SSH protocols are supported for online Git repos (e.g.
+  `[user@]server:project.git`).
+
+  A `dvc.api.UrlNotDvcRepoError` is raised if `repo` is not a valid DVC project.
+
+- `rev` - Git commit (any [revision](https://git-scm.com/docs/revisions) such as
+  a branch or tag name, or a commit hash). If not supplied, it uses the default
+  Git revision, `HEAD`. If `repo` is a Git repo, this option is ignored.
+
+- `remote` - name of the [DVC remote](/doc/command-reference/remote) to look for
+  the target data. If not supplied, the cache directory is tried first for local
+  projects; The default remote of `repo` is tried otherwise.
+
+  A `dvc.exceptions.NoRemoteError` is raised if no `remote` is found.
+
+- `mode` - (optional) mirrors the namesake parameter in builtin
+  [`open()`](https://docs.python.org/3/library/functions.html#open). Defaults to
+  `"r"` (read).
+
+- `encoding` - (optional) used to decode contents to a string. Mirrors the
+  namesake parameter in builtin `open()`. Defaults to `"utf-8"`.
+
+## Example: Process XML file from an external DVC repository
+
+```py
+from xml.dom.minidom import parse
+
+import dvc.api
+
+with dvc.api.open(
+        'get-started/data.xml',
+        repo='https://github.com/iterative/dataset-registry'
+        ) as fd:
+    xmldom = parse(fd)
+    # ... Process elements
+```
+
+> See `dvc.api.read()` for a shorthand way to read the contents of a tracked
+> file.
+
+## Example: Use a file from the local cache
+
+In this case we don't supply a `repo` value. DVC will walk up the current
+working directory tree to find the <abbr>DVC project</abbr>:
+
+```py
+import dvc.api
+
+with dvc.api.open('data/nlp/words.txt') as fd:
+    for word in fd:
+        # ... Process words
+```
+
+DVC will look for `data/nlp/words.txt` in the local cache of the
+<abbr>project</abbr>. (If it's not found there, the default
+[remote](/doc/command-reference/remote) will be tried.)
+
+To specify the file encoding of a text file:
+
+```py
+with dvc.api.open('data/nlp/words.txt', encoding='utf-8') as fd:
+    # ...
+```
+
+## Example: Process CSV file from a private repository
```

Comment:
Feels like we have too many examples and they are very similar - CSV vs XML does not provide additional info to the `open` itself. We need to show the basic usage, show local repo, should using rev flag - like open a specific tag from a local repo or remote repo.

Reply:
The difference between CSV and XML is reading line by line and reading everything at once. I think the issue is that all this is just way too wordy. Should be cut 4-5 times. Like 1 sentence for a param, no types, say about corner cases. 1-2 sentence explanation for each example, do not make a whole subsection for each. Maybe even use single code block like [here](https://github.com/Suor/django-cacheops#setup), explain things in the short comments.

Reply:
We don't write a Python tutorial though. I doubt that we need to educate people in the API reference on different ways to read streams.

Reply:
> We need to show the basic usage, show local repo, should using rev flag - like open a specific tag

Done in SDFASD (only `rev` example was missing).

I agree with Ivan in that we don't need both XML and CSV examples. HOWEVER I did keep the CSV code for the missing `rev` example ^

> this is just way too wordy...

I tied to make it more compact now and include several code samples per example section. It's still a little verbose but that's just our style, at least the same as in the cmd ref.